### PR TITLE
feat(unlock-app): Add swap and purchase payment option

### PR DESCRIPTION
**Pull request by searchableguy**
# Description

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: packages/unlock-js/src/walletService.ts [hunk 1]
```diff
@@ -14,6 +16,47 @@ interface CreateLockOptions {
   creator?: string
 }
 
+export const UnlockSwapPurchaserABI = [
+  `function swapAndCall(address lock, address srcToken, uint amountInMax,address uniswapRouter, bytes memory swapCalldata, bytes memory callData) public payable returns(bytes memory)`,
+]
```
Comment (julien51) [inline packages/unlock-js/src/walletService.ts:21]: Let's move to a specific file like we do for ERC20 ABI or the Key Manager contract ABI for calrity

File: packages/unlock-js/src/walletService.ts [hunk 2]
```diff
@@ -997,4 +1013,150 @@ export default class WalletService extends UnlockService {
...
+      swapCallData,
+      callData,
+      {
+        value: amountInMax,
+      }
+    )
+    return tx
+  }
+
+  async swapAndPurchase({
```
Comment (julien51) [inline packages/unlock-js/src/walletService.ts:1065]: A lot of the `purchase` logic is reused here from the `purchase` function. Is there a way we could DRY things up a bit? I worry about diverging implementation when it is really doing mostly the same thing, no?

File: packages/unlock-js/src/walletService.ts [hunk 3]
```diff
@@ -997,4 +1013,150 @@ export default class WalletService extends UnlockService {
...
+      (km) => km || ethers.constants.Zero
+    )
+
+    const referrers = (_referrers || defaultArray).map(
+      (km) => km || ethers.constants.Zero
+    )
+
+    const data = (_data || defaultArray).map((d) => d || [])
+
+    const args = [keyPrices, owners, referrers, keyManagers, data]
```
Comment (julien51) [inline packages/unlock-js/src/walletService.ts:1121]: This ^^ can all be taken from `purchase`, no? Since this is also version specific I think it would be really nice to just extract that from the existing purchase function in each version to `buildPurchaseArgs` and just use it here?

File: packages/unlock-js/src/walletService.ts [hunk 4]
```diff
@@ -997,4 +1013,150 @@ export default class WalletService extends UnlockService {
...
+    const callData = lockContract.interface.encodeFunctionData('purchase', args)
+
+    const approvedAmount = await getAllowance(
```
Comment (julien51) [inline packages/unlock-js/src/walletService.ts:1124]: Same. ^^

File: smart-contracts/.openzeppelin/polygon-mumbai.json [hunk 5]
```diff
@@ -0,0 +1,816 @@
+{
```
Comment (julien51) [inline smart-contracts/.openzeppelin/polygon-mumbai.json:1]: Let's remove these files please

File: smart-contracts/test/helpers/uniswapV3.js [hunk 6]
```diff
@@ -1,23 +1,43 @@
-const bn = require('bignumber.js')
```
Comment (julien51) [inline smart-contracts/test/helpers/uniswapV3.js:1]: Is this just the linter? If so please let's move to another PR

File: unlock-app/src/components/interface/checkout/main/Payment.tsx [hunk 7]
```diff
@@ -66,55 +71,61 @@ export function Payment({ injectedProvider, checkoutService }: Props) {
...
-        balance,
-        networkBalance,
-        isPayable,
-        isGasPayable,
-      }
-
-      return options
-    }
-  )
+  const { isLoading: isBalanceLoading, data: balance } = useBalance({
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/main/Payment.tsx:74]: I think we can re-use this!

**Review by julien51 — CHANGES_REQUESTED** [review]
Ok that's a great start! I think we need to DRY up things in unlock-js though.

File: packages/unlock-js/src/PublicLock/utils/approveAllowance.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+import { ZERO } from '../../constants'
+import { approveTransfer, getAllowance } from '../../erc20'
+import utils from '../../utils'
+
+export default async function approveAllowance({
+  erc20Address,
+  lockAddress,
+  totalAmountToApprove,
+}) {
+  // fix ERC20 allowance
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/utils/approveAllowance.js:10]: fix?

File: packages/unlock-js/src/PublicLock/utils/approveAllowance.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+      lockAddress,
+      this.provider,
+      this.signer.getAddress()
+    )
+
+    if (
+      !approvedAmount ||
+      utils.bigNumberify(approvedAmount).lt(totalAmountToApprove)
+    ) {
+      // We must wait for the transaction to pass if we want the next one to succeed!
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/utils/approveAllowance.js:23]:
```suggestion
      // We must wait for the transaction to pass if we want estimates for the next one to succeed!
```

File: packages/unlock-js/src/PublicLock/utils/parseLockGetters.js [hunk 10]
```diff
@@ -55,20 +55,20 @@ export default async function (address, provider) {
     update.keyPrice = utils.fromWei(update.keyPrice, 'ether')
     const balance = await provider.getBalance(address)
     update.balance = utils.fromWei(balance, 'ether')
+    update.currencyDecimals = 18
   } else {
     // Otherwise need to get the erc20's decimal and convert from there, as well as the symbol
     // TODO : make these calls in parallel
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/utils/parseLockGetters.js:61]: remove :)

File: packages/unlock-js/src/PublicLock/v10/purchaseKeys.js [hunk 11]
```diff
@@ -17,106 +16,29 @@ import formatKeyPrice from '../utils/formatKeyPrice'
...
-    recurringPayments,
-    decimals,
-    totalApproval, // explicit approval amount
-    data: _data,
-  },
+  options,
   transactionOptions = {},
-  callback
+  callback,
+  swap
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/v10/purchaseKeys.js:22]: Let's move that into `options` please so we keep the same signature

File: packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  erc20Address,
+  decimals,
+  owners = [],
+  keyManagers = [],
+  keyPrices = [],
+  referrers = [],
+  data = [],
+}) {
+  // If erc20Address was not provided, get it
+  const lockContract = await this.getLockContract(lockAddress)
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js:15]: Should we do that only if `erc20Address` is undefined?

File: packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+  erc20Address = await lockContract.tokenAddress()
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js:16]: shoould we do that only if `erc20Address` is not set?

File: packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
+import formatKeyPrice from './formatKeyPrice'
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js:1]: Since this is versioned (there is a v10 version) can we please just use a version for this rather than in the `utils` folder? This way we remain consistent and we can cleanup more easily.

File: packages/unlock-js/src/PublicLock/v4/index.js [hunk 15]
```diff
@@ -15,8 +15,10 @@ import updateLockName from './updateLockName'
 import updateLockSymbol from './updateLockSymbol'
 import setBaseTokenURI from './setBaseTokenURI'
 import updateRefundPenalty from './updateRefundPenalty'
+import getPurchaseKeysArguments from '../utils/getPurchaseKeysArguments'
```
Comment (julien51) [inline packages/unlock-js/src/PublicLock/v4/index.js:18]: just move `getPurchaseKeysArguments` into this directory :)

File: packages/unlock-js/src/UnlockSwapPurchaserABI.ts [hunk 16]
```diff
@@ -0,0 +1,3 @@
+export const UnlockSwapPurchaserABI = [
```
Comment (julien51) [inline packages/unlock-js/src/UnlockSwapPurchaserABI.ts:1]: Please check where/if we already have the ABI for ERC20 and use the same folder?

File: packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js (hunk 13 above)
Comment (searchableguy) [reply packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js:16]: It's just getting an erc20address from lock if none provided. It's to avoid overfetching if user already got erc20 for lock.

File: packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js (hunk 14 above)
Comment (searchableguy) [reply packages/unlock-js/src/PublicLock/utils/getPurchaseKeysArguments.js:1]: It is not versioned.

File: unlock-app/src/components/interface/checkout/main/Payment.tsx [hunk 17]
```diff
@@ -33,29 +37,33 @@ interface AmountBadgeProps {
...
+      <CryptoIcon size={16} symbol={symbol} />
     </div>
   )
 }
 
 export function Payment({ injectedProvider, checkoutService }: Props) {
   const [state, send] = useActor(checkoutService)
   const config = useConfig()
-  const { quantity, recipients } = state.context
+  const { recipients } = state.context
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/main/Payment.tsx:49]: Are we not using quantity anymore here?

**Comment by julien51** [comment]
Please add the ability to swap a native coin into the lock's currency please!

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx [hunk 18]
```diff
@@ -113,6 +113,11 @@ export function Confirm({
     keyPrice,
```
[the remark this answers is not in the document]
Comment (searchableguy) [reply unlock-app/src/components/interface/checkout/main/Confirm.tsx:113]: I will do this after this. It got too big.

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx [hunk 19]
```diff
@@ -349,6 +359,53 @@ export function Confirm({
...
+              amountInMax: ethers.utils
+                .parseUnits(
+                  payment.route
+                    .convertToQuoteToken(pricingData!.total.toString())
+                    .toFixed(4),
+                  payment.route.quote.currency.decimals
+                )
+                // 1% slippage buffer
+                .mul(101)
+                .div(100)
```
Comment (searchableguy) [inline unlock-app/src/components/interface/checkout/main/Confirm.tsx:404]: Added this since sometimes it was undervaluing. Maybe it could be lower like 0.5%?

Comment (julien51) [reply unlock-app/src/components/interface/checkout/main/Confirm.tsx:404]: I think we should optimize for the tx passing, not for the best price. Given the small amounts, even 3% is better than a failed tx (because the gas would be wasted!)

File: unlock-app/src/hooks/useUniswapRoutes.ts [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+    ) {
+      routes.push({
+        tokenIn: nativeOnChain(lock.network),
+        tokenOut,
+        amountOut: price,
+        recipient,
+        network,
+      })
+    }
+    console.log(routes)
```
Comment (julien51) [inline unlock-app/src/hooks/useUniswapRoutes.ts:123]:
```suggestion
```

File: unlock-app/src/components/interface/checkout/main/Confirm.tsx (hunk 19 above)
Comment (searchableguy) [reply unlock-app/src/components/interface/checkout/main/Confirm.tsx:411]: ok, 1% is fine for now. Works for small amounts since that's what I'm testing with.

File: unlock-app/src/hooks/useUniswapRoutes.ts [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+  routes: UniswapRoute[]
+  enabled?: boolean
+}
+
+export const useUniswapRoutes = ({
+  routes,
+  enabled = true,
+}: UniswapRoutesOption) => {
+  const web3Service = useWeb3Service()
+  console.log(routes)
```
Comment (julien51) [inline unlock-app/src/hooks/useUniswapRoutes.ts:28]:
```suggestion
```

**Review by julien51 — APPROVED** [review]

File: unlock-app/src/hooks/useUniswapRoutes.ts [hunk 22]
```diff
@@ -24,22 +26,34 @@ export const useUniswapRoutes = ({
...
             })
+            const balance = await getAccountTokenBalance(
+              web3Service,
+              account!,
+              route.tokenIn instanceof Token ? route.tokenIn.address : null,
+              route.network
+            )
+            // If the balance is less than the quote, we cannot make the swap.
+            if (parseFloat(balance) < parseFloat(response.quote.toFixed(2))) {
+              throw new Error('Insufficient balance')
```
Comment (julien51) [inline unlock-app/src/hooks/useUniswapRoutes.ts:55]: Maybe no need to throw but just console.log?

**Review by julien51 — APPROVED** [review]

File: unlock-app/src/hooks/useUniswapRoutes.ts (hunk 22 above)
Comment (searchableguy) [reply unlock-app/src/hooks/useUniswapRoutes.ts:55]: it is inside a try catch so does log and filter the route.
